### ISSUE TITLE
'git gc' hangs/crashes when working with geany

### ISSUE THREAD

**Issue by ascherer** [git-for-windows/git#1068]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.11.1.windows.1
built from commit: 1c1842bcba45569a84112ec64f72b08eb2d57c68
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation. Alle Rechte vorbehalten.
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

cat: /etc/install-options.txt: No such file or directory
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

In older versions of Git for Windows, it was possible to abort `git gc` with `Ctrl-C`.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
cd github/debbuild
[Edit some files of this project in Geany 1.29]
git gc
```
 - What did you expect to occur after running these commands?

I expected to see `git gc` run to completion and compactify `.git/objects/`.

 - What actually happened instead?

`git gc` hangs. When I _quit_ Geany 1.29, I can `Ctrl-C` the running `git gc`. Trying `git gc` again results in the warning message
```
$ git gc
fatal: gc is already running on machine 'WKM-W040-80XA' pid 6892 (use --force if not)
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

The problem is reproducible with _any_ local repository.

**Comment by dscho** [comment]
It may be possible that `git gc` wanted to print an error that some process is blocking some file or directory from being removed. A bug in v2.11.1 would prevent that error from *actually* be printed (the buggy code waits until more stuff accumulates to be printed, but that does not happen, of course).

Would you mind testing this hypothesis by installing, say, https://github.com/git-for-windows/git/releases/tag/v2.11.1.windows-prerelease.2?

**Comment by ascherer** [comment]
```
git version 2.11.1.windows.prerelease.2
built from commit: 86d49ccac7afb39e03ba2632803c5ab67d4bb792
sizeof-long: 4
machine: x86_64
```
shows the same behavior as before. No error/warning message is printed.

**Comment by dscho** [comment]
> 86d49ccac7afb39e03ba2632803c5ab67d4bb792

Oh bummer, sorry, I forgot that this version *already* had the "buffered stderr" bug. I *think* the version to verify that there is a question hanging is v2.10.2. For testing, I would suggest the portable version: https://github.com/git-for-windows/git/releases/download/v2.10.2.windows.1/PortableGit-2.10.2-64-bit.7z.exe

**Comment by ascherer** [comment]
```
git version 2.10.2.windows.1
sizeof-long: 4
machine: x86_64
```
correctly reports
```
$ git gc
Counting objects: 557, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (185/185), done.
Writing objects: 100% (557/557), done.
Total 557 (delta 368), reused 557 (delta 368)
Unlink of file '.git/objects/pack/pack-b9c6732d813db7d9882d43795c912415a84936fb.pack' failed. Should I try again? (y/n)
```
If I quit Geany and answer `y`, `git gc` runs to completion. (If I _abort_ with `Ctrl-C` and rerun, I am requested to use option `--force`.)

**Comment by dscho** [comment]
Okay, good. It would appear that [the bug fix](9dc4268eeebfac6e3437fec879fca91361aba297) we already have in `master` fixes this issue, too, then.

Even if it is not obvious from the descriptions, the same bug is the reason for https://github.com/git-for-windows/git/issues/1062 and https://github.com/git-for-windows/git/issues/1064.
